### PR TITLE
863 fix chado_db_select function

### PR DESCRIPTION
**Pull request by bradfordcondon**
# Bug Fix

Issue #863 

## Description
The `chado_db_select`  API function was incompatible with db transactions.  This was a problem for both tests and importers.  Furthermore, it only resolved `select` statements, and not joins.  It also didnt allow joining across public/chado schemas.

  To resolve, the transaction issue, we no longer create a new connection.  In addition, we utilize a [Drupal Query Extender](https://www.drupal.org/docs/7/api/database-api/dynamic-queries/extenders) to modify the query returned by `chado_db_select`.

## Testing?
I've added a unit test that demonstrates that the problem (`test_chado_db_select`)- the query with `chado_db_select` fails because the test takes place in a transaction, but the identical query succeeds.  On this branch, the test will pass.    

## Additional Notes (if any):
This PR was co-authored with @almasaeed2010

### CODE REVIEW

**Comment by bradfordcondon** [comment]
the breaking test made us realize that `chado_query` and `chado_db_select` behave differently.  We should [discuss what we want to do before merging](https://github.com/tripal/tripal/issues/863#issuecomment-468747801).

**Comment by almasaeed2010** [comment]
I think this is ready. [The new extender class](https://github.com/tripal/tripal/pull/867/files#diff-2afcc316886a934e8cf98c911333ddbb) no longer establishes a new connection and instead overrides  the select  and the join (and its variants) functions of the SelectQuery class to prefix chado tables with the correct schema. It also looks for `public.` and replaces it with the proper schema name if different.

**Comment by bradfordcondon** [comment]
pointing this out because its hard to distinguish failing tests from the current NCBI-related build error:

Looks like `test_chado_db_select_recognizes_non_chado_tables` is failing on Travis
```

Time: 4.14 minutes, Memory: 60.00 MB
There was 1 error:
1) Tests\ChadoQueryTest::test_chado_db_select_recognizes_non_chado_tables
PDOException: SQLSTATE[42601]: Syntax error: 7 ERROR:  syntax error at or near "FROM"
LINE 2: FROM 
        ^
/home/travis/build/tripal/drupal/includes/database/database.inc:2227
/home/travis/build/tripal/drupal/includes/database/pgsql/database.inc:111
/home/travis/build/tripal/drupal/includes/database/select.inc:1280
/home/travis/build/tripal/drupal/includes/database/select.inc:724
/home/travis/build/tripal/drupal/sites/all/modules/tripal/tests/tripal_chado/api/ChadoQueryTest.php:108

```

**Comment by almasaeed2010** [comment]
Looks like the tests are passing now. 
Thanks @bradfordcondon for the help!

I'll mark this as ready for review.

**Comment by spficklin** [comment]
@almasaeed2010 in the process of trying to fix merge conflicts with the big reformatting that was just merged into the master branch there were some new functions added that showed up as new and when asking @bradfordcondon about them on Slack he wasn't 100% sure about them.  Can you confirm that these functions: `chado_replace_schema_prefix` and `chado_replace_table_prefix` should be included in the PR?

**Comment by almasaeed2010** [comment]
Hello @spficklin. Yes both should be included although we might want to rename them.

|Function|Description|
|--------|------------|
|`chado_replace_schema_prefix`|Takes in a table name with or without schema such as `chado.analysis`, `public.users`, `users` and finds the correct name for the schema. For example, chado.analysis will be replaces with `chado_get_schema_name('chado').'analysis'`.|
|`chado_replace_table_prefix`| Takes in SQL and replaces braces with table names.|

Since the `chado_replace_schema_prefix` function is only used in the extender class, I'll move it there as a static function.

**Comment by almasaeed2010** [comment]
Moved function to the extender class as `static::getRealSchema()` and added unit tests for it.

**Comment by spficklin** [comment]
Thanks @almasaeed2010. Do you have a suggestion for renaming of those?  Once they are in the API it's really hard to change names.  So, we should do that now rather than later.

**Comment by almasaeed2010** [comment]
The `chado_replace_schema_prefix` is not part of the extender so we don't need to worry about that. And I think `chado_replace_table_prefix` can be left as is since it's no longer confusing.

**Comment by bradfordcondon** [comment]
@almasaeed2010 if these were in a class both of those functions would be `private` right? 

Can we can move them out of the `.api` file and/or prefix them with `_` to get around this issue?


`chado_replace_schema_prefix` I no longer see this in the code.  however its still referenced in `getTable` so we should remove that
`chado_replace_table_prefix` is only used 1x so it could be put directly in the code (which was how it started!) or prefixed with `_` to make clear its not part of the API...

File: tripal_chado/includes/ChadoPrefixExtender.inc
Comment (bradfordcondon) [inline tripal_chado/includes/ChadoPrefixExtender.inc]: This function was removed
```suggestion
```

**Comment by almasaeed2010** [comment]
To clarify, `chado_replace_schema_prefix` no longer exists. And I just removed `chado_replace_table_prefix` and used the code directly in chado_query as it was before so it's no longer an API function.

Thanks!

**Review by spficklin — APPROVED** [review]

**Comment by spficklin** [comment]
After having looked at the code changes and testing I see no problems with this fix.  I approve.

**Comment by bradfordcondon** [comment]
grateful to Stephen for the review.  Is anyone else able to review this? It's blocking me from submitting Tripal HQ as a gold standard module. Apologies for the pings, know we're busy.  @laceysanderson @shawnawsu @guignonv @Ferrisx4 @chunhuaicheng

**Comment by guignonv** [comment]
@bradfordcondon: Too busy this week unfortunately but if nothing is done by then, I'll check/validate the changes next week.

**Comment by guignonv** [comment]
I've created a new dev website to test this. I think the changes may have a deep impact on how the multi-chado module works so I'll have to dig a bit and test a couple of things. I think we should also reference issue #431 as it is related.
I'll do some tests and come back to you.

I'll just add a point about the Multi-Chado module: to trick Drupal/Tripal and handle different linking tables between Chado instances, it dynamically replace the linking tables (chado_XXX for Tripal 2.x and chado_bio_data_XXX for Tripal 3.x) with a temporary view which will "redirect" to the appropriate tables corresponding to the right chado schema. I know it's a bit short to explain how it works and to understand... but basically, it means that if a "public." prefix is hardcoded in queries (and there's no hook to remove it), the multi-chado won't work.

**Edit**: just to add more information about how postgres handles schema, when querying a table, postgres not only uses the "search_path" but first look at the temporary objects. So if you do `SELECT * FROM foo;` and uses a search_path like `"$user", public, chado`, postgresql will first query the temporary table/view `foo`, then the one from the user schema, then the one from the public schema and finally the one from the chado schema. However, if your query is `SELECT * FROM public.foo;`, there's no ambiguity and postgres will use the public schema without taking into account any temporary table.

**Comment by spficklin** [comment]
Hi @guignonv good catch on the hard-coded `public` and `chado` schemas. I fixed those and just pushed to this same branch. They were hardcoded in the `getTable` function.  But all was good in the `getRealSchema` function.

Can you review again please?  If you don't have time to do a full functional test, as long as you are happy with a visual code review I think we can merge, as I've already tested it.

To keep our PRs more simple I think we should fix issue #431 in a different PR rather than try to fit it into this one. I'm not sure if that's what you meant when you referred to it.

**Comment by guignonv** [comment]
I have quite a by stack of tasks this week. I'll try to find a moment next week.

**Comment by laceysanderson** [comment]
A note regarding **"Some checks were not successful"**. As you can see [here](https://travis-ci.org/tripal/tripal/jobs/524709325#L991) all tests against the 7.x-3.x branch pass ✅ . The failure is happening during the [tripal 2 upgrade tests](https://travis-ci.org/tripal/tripal/jobs/524709325#L1117) which is a problem already solved in previous PRs.

**Review by laceysanderson — APPROVED** [review]
I did both a code and functional test and it all looks good 👍 

Specifically, I tested 
- [x] a number of different fetch methods
- [x] queried a non-standard chado table (phenotype_ to ensure additional columns showed up and that I could filter by them
- [x] queried a chado custom table (genotype_call) and joined on it to a standard chado table (project)
- [x] Union between feature and stock selecting the name, uniquename

**Comment by laceysanderson** [comment]
Since this has two approvals as required by our guidelines, I'm merging it. If you run into any problems with it @guignonv please comment on the associated issue!

**Pull request merged**
